### PR TITLE
feat: add VaadinService scope

### PR DESCRIPTION
**Pull request by denis-anisimov**
## Description

Please list all relevant dependencies in this section and provide summary of the change, motivation and context.

Fixes # (issue)

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [ ] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [ ] I have added a description following the guideline.
- [ ] The issue is created in the corresponding repository and I have referenced it.
- [ ] I have added tests to ensure my change is effective and works as intended.
- [ ] New and existing tests are passing locally with my change.
- [ ] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: runtime/src/main/java/com/vaadin/quarkus/context/BeanProvider.java [hunk 1]
```diff
@@ -0,0 +1,597 @@
...
+     * 
+     * private static <T> DependentProvider<T> createDependentProvider(
+     * BeanManager beanManager, Class<T> type, Bean<T> bean) {
+     * CreationalContext<T> cc = beanManager.createCreationalContext(bean);
+     * 
+     * @SuppressWarnings("unchecked") T instance = (T)
+     * beanManager.getReference(bean, type, cc);
+     * 
+     * return new DependentProvider<T>(bean, cc, instance); }
+     */
```
Comment (caalador) [inline runtime/src/main/java/com/vaadin/quarkus/context/BeanProvider.java:444]: Commented out method should be uncommented or removed.

File: runtime/src/test/java/com/vaadin/quarkus/TestQuarkusVaadinServletService.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        return false;
+    }
+
+    @Override
+    public void setClassLoader(ClassLoader classLoader) {
+        if (classLoader != null) {
+            super.setClassLoader(classLoader);
+        }
+    }
+}
```
Comment (caalador) [inline runtime/src/test/java/com/vaadin/quarkus/TestQuarkusVaadinServletService.java:64]: new line

File: runtime/src/test/java/com/vaadin/quarkus/context/ServiceContextTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    protected Class<VaadinServiceScopedContext> getContextType() {
+        return VaadinServiceScopedContext.class;
+    }
+
+    @VaadinServiceScoped
+    public static class ServiceScopedTestBean extends TestBean {
+    }
+
+}
```
Comment (caalador) [inline runtime/src/test/java/com/vaadin/quarkus/context/ServiceContextTest.java:56]: new line

File: runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+import org.mockito.Mockito;
+
+import com.vaadin.flow.internal.ReflectTools;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertSame;
+
+public abstract class AbstractContextTest<T extends TestBean, C extends AbstractContext> {
```
Comment (caalador) [inline runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java:43]: Could perhaps have a javadoc mentioning that this has the basic tests for all custom contexts.

File: runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+        return underTestContext;
+    }
+
+    protected abstract UnderTestContext newContextUnderTest();
+
+    protected abstract boolean isNormalScoped();
+
+    protected abstract Class<T> getBeanType();
+
+    protected abstract Class<C> getContextType();
```
Comment (caalador) [inline runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java:246]: All the abstracts should have a javadoc with a minimal on what is expected of the implementation

File: integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/SmokeTest.java [hunk 6]
```diff
@@ -26,7 +25,8 @@ public void smokeTest_clickButton() {
 
         button.click();
 
-        // As we don't have CDI functionality for Route annotations we check that we are in
+        // As we don't have CDI functionality for Route annotations we check
+        // that we are in
```
Comment (caalador) [inline integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/SmokeTest.java:29]: This comment is actually wrong and should be removed.

File: integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/AbstractCdiTest.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        slurp("?resetCounts");
+    }
+
+    protected int getCount(String id) throws IOException {
+        getCommandExecutor().waitForVaadin();
+        String line = slurp("?getCount=" + id);
+        return Integer.parseInt(line);
+    }
+
+    private String slurp(String uri) throws IOException {
```
Comment (caalador) [inline integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/AbstractCdiTest.java:63]: slurp? This naming is from clojure? also slurp should read the whole file, but in this instance we only read the first line in the response. Also it doesn't take a uri but queryParameters for the CounterFilter.

 Perhaps another name would be more appropriate to lessen confusion on what it does, like `doAction(String parameter)` or something.

**Review by caalador — CHANGES_REQUESTED** [review]

File: runtime/src/main/java/com/vaadin/quarkus/AnyLiteral.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.quarkus;
+
+import javax.enterprise.inject.Any;
+import javax.enterprise.util.AnnotationLiteral;
+
+public class AnyLiteral extends AnnotationLiteral<Any> implements Any {
+}
```
[the remark this answers is not in the document]
Comment (denis-anisimov) [reply runtime/src/main/java/com/vaadin/quarkus/AnyLiteral.java:22]: done

File: runtime/src/main/java/com/vaadin/quarkus/AnyLiteral.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.quarkus;
+
+import javax.enterprise.inject.Any;
+import javax.enterprise.util.AnnotationLiteral;
+
+public class AnyLiteral extends AnnotationLiteral<Any> implements Any {
```
[the remark this answers is not in the document]
Comment (denis-anisimov) [reply runtime/src/main/java/com/vaadin/quarkus/AnyLiteral.java:21]: done

File: runtime/src/main/java/com/vaadin/quarkus/context/BeanProvider.java (hunk 1 above)
Comment (denis-anisimov) [reply runtime/src/main/java/com/vaadin/quarkus/context/BeanProvider.java:444]: done

File: runtime/src/test/java/com/vaadin/quarkus/TestQuarkusVaadinServletService.java (hunk 2 above)
Comment (denis-anisimov) [reply runtime/src/test/java/com/vaadin/quarkus/TestQuarkusVaadinServletService.java:64]: done

File: runtime/src/test/java/com/vaadin/quarkus/context/ServiceContextTest.java (hunk 3 above)
Comment (denis-anisimov) [reply runtime/src/test/java/com/vaadin/quarkus/context/ServiceContextTest.java:56]: done

File: runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java (hunk 4 above)
Comment (denis-anisimov) [reply runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java:43]: done

File: runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java (hunk 5 above)
Comment (denis-anisimov) [reply runtime/src/test/java/com/vaadin/quarkus/context/AbstractContextTest.java:246]: this is a test class which is modified copy of https://github.com/vaadin/cdi/blob/master/vaadin-cdi/src/test/java/com/vaadin/cdi/context/AbstractContextTest.java.
The latter class has no any javadocs.

File: integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/SmokeTest.java (hunk 6 above)
Comment (denis-anisimov) [reply integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/SmokeTest.java:29]: done. Even though it's out of scope of the PR and formatting is done automatically.

File: integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/AbstractCdiTest.java (hunk 7 above)
Comment (denis-anisimov) [reply integration-tests/common-test-code/src/test/java/com/vaadin/flow/quarkus/it/AbstractCdiTest.java:63]: The method comes from https://github.com/vaadin/cdi/blob/master/vaadin-cdi-itest/src/test/java/com/vaadin/cdi/itest/AbstractCdiTest.java.
This is just a copy .
I don't want to rename the method even though I don't like the name. It's better to keep the consistency between original code and copy.

**Review by caalador — APPROVED** [review]

**Pull request merged**